File: src/knowledge-ingestion/chunker.ts
import { CanonicalDocument, DocumentChunk, DocumentSection } from '../shared/types';

export class DocumentChunker {
  private targetSize: number; // target size in characters (e.g., 1000 chars is ~200-250 words)
  private overlap: number;

  constructor(targetSize = 1000, overlap = 150) {
    this.targetSize = targetSize;
    this.overlap = overlap;
  }

  public chunk(doc: CanonicalDocument): Omit<DocumentChunk, 'id'>[] {
    const chunks: Omit<DocumentChunk, 'id'>[] = [];
    let chunkIndex = 0;

    const enrichContent = (docName: string, path: string[], text: string) => {
      const pathHeader = path.length > 0 ? ` > ${path.join(' > ')}` : '';
      return `[Source: ${docName}${pathHeader}]\n\n${text}`;
    };

    // Helper to walk the section tree recursively
    const walk = (
      sections: DocumentSection[],
      currentPath: string[],
      metadata: Record<string, any>
    ) => {
      let buffer = '';
      let bufferMetadata: Record<string, any> = { ...metadata };

      const flushBuffer = () => {
        if (!buffer.trim()) return;
        const enriched = enrichContent(doc.name, currentPath, buffer.trim());
        chunks.push({
          workspaceId: doc.workspaceId,
          documentId: doc.documentId,
          documentName: doc.name,
          sectionPath: [...currentPath],
          chunkIndex: chunkIndex++,
          content: enriched,
          metadata: {
            sourceType: doc.fileType,
            ...bufferMetadata
          }
        });
        buffer = '';
      };

      for (const section of sections) {
        // Build the section path
        const sectionPath = section.title 
          ? [...currentPath, section.title] 
          : currentPath;

        // If the section has content, process it
        if (section.content) {
          const sectionHeader = section.title ? `### ${section.title}\n` : '';
          const contentToAppend = sectionHeader + section.content;

          // If adding this content exceeds target size, flush first
          if (buffer.length + contentToAppend.length > this.targetSize && buffer.length > 0) {
            flushBuffer();
          }

          // Accumulate metadata from the section (e.g. page, sheet, row)
          if (section.source) {
            bufferMetadata = { ...bufferMetadata, ...section.source };
          }

          buffer += (buffer ? '\n\n' : '') + contentToAppend;
        }

        // If it has children, walk them
        if (section.children && section.children.length > 0) {
          // If buffer is already near target, flush it first to keep child sections separate
          if (buffer.length > this.targetSize * 0.7) {
            flushBuffer();
          }
          
          walk(section.children, sectionPath, { ...bufferMetadata });
        } else {
          // If it's a leaf node paragraph and it's huge, we check if we need to split it
          if (section.content && section.content.length > this.targetSize) {
            flushBuffer(); // Flush anything prior
            
            // Chunk huge paragraphs by sentences
            const sentences = section.content.match(/[^.!?]+[.!?]+(\s|$)/g) || [section.content];
            let subBuffer = '';
            
            for (const sentence of sentences) {
              if (subBuffer.length + sentence.length > this.targetSize && subBuffer.length > 0) {
                const enriched = enrichContent(doc.name, sectionPath, subBuffer.trim());
                chunks.push({
                  workspaceId: doc.workspaceId,
                  documentId: doc.documentId,
                  documentName: doc.name,
                  sectionPath: [...sectionPath],
                  chunkIndex: chunkIndex++,
                  content: enriched,
                  metadata: {
                    sourceType: doc.fileType,
                    ...section.source
                  }
                });
                // Carry over overlap
                subBuffer = subBuffer.slice(-this.overlap);
              }
              subBuffer += sentence;
            }
            
            if (subBuffer.trim()) {
              buffer = subBuffer;
              bufferMetadata = { ...bufferMetadata, ...section.source };
            }
          }
        }
      }

      // Flush remaining content in buffer
      flushBuffer();
    };

    walk(doc.sections, [], {});
    return chunks;
  }
}

File: backend/app/services/ingestion/chunker.py
from __future__ import annotations

import re
from dataclasses import dataclass

from .parsers import CanonicalDocument, DocumentChunk, DocumentSection


@dataclass
class ChunkerConfig:
    target_size: int = 1000
    overlap: int = 150


class DocumentChunker:
    def __init__(self, config: ChunkerConfig | None = None):
        self.config = config or ChunkerConfig()

    def chunk(self, doc: CanonicalDocument) -> list[DocumentChunk]:
        chunks: list[DocumentChunk] = []
        chunk_index = 0

        def enrich_content(path: list[str], text: str) -> str:
            path_header = f" > {' > '.join(path)}" if path else ""
            return f"[Source: {doc.name}{path_header}]\n\n{text}"

        def walk(sections: list[DocumentSection], current_path: list[str], metadata: dict) -> None:
            nonlocal chunk_index
            buffer = ""
            buffer_metadata = dict(metadata)

            def flush_buffer() -> None:
                nonlocal buffer, chunk_index
                if not buffer.strip():
                    return
                chunks.append(
                    DocumentChunk(
                        workspace_id=doc.workspace_id,
                        document_id=doc.document_id,
                        document_name=doc.name,
                        section_path=list(current_path),
                        chunk_index=chunk_index,
                        content=enrich_content(current_path, buffer.strip()),
                        metadata={"sourceType": doc.file_type, **buffer_metadata},
                    )
                )
                chunk_index += 1
                buffer = ""

            for section in sections:
                section_path = [*current_path, section.title] if section.title else list(current_path)
                if section.content:
                    section_header = f"### {section.title}\n" if section.title else ""
                    content_to_append = section_header + section.content
                    if len(buffer) + len(content_to_append) > self.config.target_size and buffer:
                        flush_buffer()
                    if section.source:
                        buffer_metadata = {**buffer_metadata, **section.source}
                    buffer += ("\n\n" if buffer else "") + content_to_append

                if section.children:
                    if len(buffer) > self.config.target_size * 0.7:
                        flush_buffer()
                    walk(section.children, section_path, dict(buffer_metadata))
                elif section.content and len(section.content) > self.config.target_size:
                    flush_buffer()
                    sentences = re.findall(r"[^.!?]+[.!?]+(?:\s|$)", section.content) or [section.content]
                    sub_buffer = ""
                    for sentence in sentences:
                        if len(sub_buffer) + len(sentence) > self.config.target_size and sub_buffer:
                            chunks.append(
                                DocumentChunk(
                                    workspace_id=doc.workspace_id,
                                    document_id=doc.document_id,
                                    document_name=doc.name,
                                    section_path=list(section_path),
                                    chunk_index=chunk_index,
                                    content=enrich_content(section_path, sub_buffer.strip()),
                                    metadata={"sourceType": doc.file_type, **section.source},
                                )
                            )
                            chunk_index += 1
                            sub_buffer = sub_buffer[-self.config.overlap :]
                        sub_buffer += sentence
                    if sub_buffer.strip():
                        buffer = sub_buffer
                        buffer_metadata = {**buffer_metadata, **section.source}

            flush_buffer()

        walk(doc.sections, [], {})
        return chunks
